### PR TITLE
chore: go tools: use AddDatasetElements

### PR DESCRIPTION
**Pull request by g-linville**
This should improve performance, since we're doing just one request and one tool call to add all the elements, instead of one per element.

### CODE REVIEW

File: outlook/calendar/pkg/commands/listEvents.go [hunk 1]
```diff
@@ -44,35 +43,42 @@ func ListEvents(ctx context.Context, start, end time.Time) error {
...
+			return fmt.Errorf("failed to create dataset: %w", err)
+		}
 
-						return fmt.Errorf("failed to add event to dataset: %w", err)
-					}
+		names := map[string]any{} // Store names to avoid duplicates
+		var elements []gptscript.DatasetElement
+		for cal, events := range calendarEvents {
+			for _, event := range events {
+				name := util.Deref(event.GetSubject()) + "_" + util.Deref(cal.Calendar.GetName()) + "_" + util.Deref(event.GetStart().GetDateTime())
```
Comment (njhale) [inline outlook/calendar/pkg/commands/listEvents.go:61]: I just checked and it is possible to create more than one distinct event in a calendar with the same subject and start time (and end time, content, etc too). I think we need to keep all the events, regardless of whether they share the same user-provided fields. There should be a unique ID field on these objects we can add to the name to do that.

Comment (g-linville) [reply outlook/calendar/pkg/commands/listEvents.go:61]: Oh, thanks. I'll fix this.

[the remark this answers is not in the document]
Comment (g-linville) [reply outlook/calendar/pkg/commands/listEvents.go]: fixed

File: outlook/calendar/pkg/commands/listEvents.go [hunk 2]
```diff
@@ -44,35 +43,42 @@ func ListEvents(ctx context.Context, start, end time.Time) error {
...
+					elements = append(elements, gptscript.DatasetElement{
+						DatasetElementMeta: gptscript.DatasetElementMeta{
+							Name:        name,
+							Description: util.Deref(event.GetBodyPreview()),
+						},
+						Contents: printers.EventToString(ctx, c, cal, event),
+					})
+					names[name] = struct{}{}
 				}
+
```
[the remark this answers is not in the document]
Comment (g-linville) [reply outlook/calendar/pkg/commands/listEvents.go]: this code has been removed

File: outlook/calendar/pkg/commands/listEvents.go (hunk 1 above)
Comment (g-linville) [reply outlook/calendar/pkg/commands/listEvents.go]: fixed

File: outlook/mail/pkg/commands/listMailFolders.go [hunk 3]
```diff
@@ -30,23 +30,35 @@ func ListMailFolders(ctx context.Context) error {
...
-				if err != nil {
-					return fmt.Errorf("failed to convert mail folder to string: %w", err)
-				}
-
-				if _, err = gptscriptClient.AddDatasetElement(ctx, os.Getenv("GPTSCRIPT_WORKSPACE_DIR"), dataset.ID, util.Deref(folder.GetId()), util.Deref(folder.GetDisplayName()), folderStr); err != nil {
-					return fmt.Errorf("failed to add element: %w", err)
-				}
+		workspaceID := os.Getenv("GPTSCRIPT_WORKSPACE_ID")
+
+		dataset, err := gptscriptClient.CreateDataset(ctx, workspaceID, "outlook_mail_folders", "Outlook mail folders")
```
Comment (thedadams) [inline outlook/mail/pkg/commands/listMailFolders.go:35]: nitL is there a reason this one has `_` and some of the others don't?

**Review by thedadams — APPROVED** [review]

File: outlook/mail/pkg/commands/listMailFolders.go (hunk 3 above)
Comment (g-linville) [reply outlook/mail/pkg/commands/listMailFolders.go:35]: nope not really

File: outlook/mail/pkg/commands/listMessages.go [hunk 4]
```diff
@@ -43,23 +43,35 @@ func ListMessages(ctx context.Context, folderID string) error {
...
-				messageStr, err := printers.MessageToString(message, false)
-				if err != nil {
-					return fmt.Errorf("failed to convert message to string: %w", err)
-				}
+		workspaceID := os.Getenv("GPTSCRIPT_WORKSPACE_ID")
 
-				if _, err = gptscriptClient.AddDatasetElement(ctx, os.Getenv("GPTSCRIPT_WORKSPACE_DIR"), dataset.ID, util.Deref(message.GetId()), util.Deref(message.GetSubject()), messageStr); err != nil {
-					return fmt.Errorf("failed to add element: %w", err)
-				}
+		dataset, err := gptscriptClient.CreateDataset(ctx, workspaceID, fmt.Sprintf("%s_outlook_mail", folderID), "Outlook mail messages in folder "+folderID)
```
Comment (drpebcak) [inline outlook/mail/pkg/commands/listMessages.go:48]: noticed same thing here as @thedadams's comment, why `_` here but not other places?

**Review by drpebcak — APPROVED** [review]

**Pull request merged**